server/hawt-auth.js

```diff
@@ -0,0 +1,211 @@
+ *  and inspired by Hawk.  Requests are authenticated by icluding the JWT
+ *  directly in the authorization header like so:
+ *
+ *    Authorization: Hawt [JWT]
+ *
+ *  To be accepted, the JWT must be signed by one of a configurable list
+ *  of trusted JKUs, and must reference the relevant JKU in its header.
+ *  The JWT claims must include the following meta-data about the request in
+ *  order to guard against reply etc:
+ *
+ *    - iss:  hostname of the service issuing the request
+ *    - aud:  must match publicly-visible domain of the target server
+ *    - exp:  must be a timestamp within configurable freshness window
+ *    - iat:  if present, must be a timestamp within configurable window
+ *    - nce:  nonce; a random string unique to this request
+ *    - qsh:  query hash; a digest of the request parameters (see below)
+ *    - psh:  payload hash; an optional hash of the request payload (see below)
+ *
+ *  The payload hash is analogous to that used in Hawk, and is the base64
+ *  encoded SHA256 digest of the concatenation of the following strings,
+ *  each followed by a newline:
+ *
+ *     - "hawt.1.payload"
+ *     - the content-type header, in lowercase, without parameters
+ *     - the request payload prior to any content encoding
+ *
+ *  The query hash is analogous to the "mac" field used in Hawk, but is a raw
+ *  hash rather than a HMAC and does not include any fields over which the JWT
+ *  already provides authentication.  It is the base64 encoded SHA256 digest
+ *  of the concatenation of the following strings, each followed by a newline:
+ *
+ *     - "hawt.1.query"
+ *     - the nonce value from the JWT claims
+ *     - the request method, in uppercase
+ *     - the request path, including query string, without normalization
+ *     - the payload hash, or an empty string if there's no payload hash
+ *
+ *  Notes and thoughts:
+ *
+ *     - do we need a `typ` claim, or is it unnecessary given `qsh`?
+ *     - make sure we reject { alg: "none" } JWTs!
+ */
+
+var crypto = require('crypto')
+var url = require('url')
+var jws = require('jws')
+var b64urltohex = require('browserid-crypto/lib/utils').b64urltohex
+var P = require('../promise')
+
+var boom = require('boom')
+
+module.exports = function (error, config) {
+
+  var jwks = require('../jwks')(error, config)
+
+  function Hawt(server, options, next) {
+    server.auth.scheme('hawt', function (server, options) {
+      return {
+
+        authenticate: function (request, reply) {
+          var authz = request.headers.authorization
+          if (!authz || authz.indexOf('Hawt ') !== 0) {
+            // AFAICT, we must return an instance of boom.unauthorized here
+            // to indicate "not our auth scheme" and allow other stragies to
+            // handle the auth.  Returning some other error causes it to
+            // fail out with a 401.
+            return reply(boom.unauthorized(null, 'Hawt'))
```

Comment:
yes. Its cryptic but it is the documented way to "pass"

> If the err does not include a message but does include a scheme name (e.g. Boom.unauthorized(null, 'Custom')), additional strategies will be attempted in order of preference.

https://github.com/hapijs/hapi/blob/master/API.md#serverauthschemename-scheme